Minimal client changes to support PKCE.

GinoCanessa opened the pull request:
For clarity, these are just the changes in `src`.

GinoCanessa committed:
Minimal client changes to support PKCE.

jmandel commented on src/smart.ts:
```diff
@@ -134,6 +136,60 @@ function any(tasks: Task[]): Promise<any> {
+   var codeVerifier = base64urlEncode(env, Array.from(output).map(function (num) {
+     return PKCE_CHARSET[num % PKCE_CHARSET.length];
+   }).join(''));
+   return crypto.subtle.digest('SHA-256', new TextEncoder().encode(codeVerifier)).then(function (buffer) {
```
Given http://docs.smarthealthit.org/client-js/#server-usage I'm expecting this will break when used in a sever-side context (where `crypto.subtle` isn't available).  If we're targeting recent node versions, we could rely on 

    const { subtle } = require('crypto').webcrypto;

(https://nodejs.org/api/webcrypto.html#webcrypto_web_crypto_api) as a fallback if not available in the browser.

jmandel commented on src/smart.ts:
```diff
@@ -134,6 +136,60 @@ function any(tasks: Task[]): Promise<any> {
+  * the same as regular base64 encoding.
+  * @param value string to encode
+  */
+ function base64urlEncode(env: fhirclient.Adapter, value:string) {
```
(For asymmetric client authn, we'll definitely want to bring in an external dependency like https://github.com/cisco/node-jose -- at which point we can rely on their `jose.util.base64url.encode` rather than implementing here.)

jmandel commented on src/smart.ts:
```diff
@@ -134,6 +136,60 @@ function any(tasks: Task[]): Promise<any> {
+   var output = new Uint32Array(RECOMMENDED_CODE_VERIFIER_LENGTH);
+   crypto.getRandomValues(output);
+   var codeVerifier = base64urlEncode(env, Array.from(output).map(function (num) {
+     return PKCE_CHARSET[num % PKCE_CHARSET.length];
```
Again assuming we need a dependency like https://github.com/cisco/node-jose anyway, we can use `jose.util.randomBytes(RECOMMENDED_CODE_VERIFIER_LENGTH)` and pass the digest into `jose.util.base64url.encode`.

GinoCanessa commented:
That works for me.  If that dependency is being added anyway then I'll happily use it here.  Is it ok to use it directly, or do you think we need to wrap add calls to `lib.ts`, so that we can have it optionally fallback to the browser `crypto` implementation?

jmandel commented:
> Is it ok to use it directly, or do you think we need to wrap add calls to lib.ts, so that we can have it optionally fallback to the browser crypto implementation?

I _think_ the cisco library will user either browser crypto or Node crypto under the hood, depending on what's available

GinoCanessa commented:
> I _think_ the cisco library will user either browser crypto or Node crypto under the hood, depending on what's available

Yes, it appears to.

GinoCanessa committed:
Added node-jose.
Replaced browser.crypto with node-jose.
New PKCE usage flagging (optional by default, can require or disable).
Added tests for PKCE.

GinoCanessa commented:
Ok, new version includes the Cisco library (node-jose), updated PKCE configuration (optional but on by default, can set to required or disabled), and I've added *some* test coverage.

I didn't see a good place to add in tests for adding the `code_verifier` to a request - if there's a place where it's mocked out properly to enable that I'm happy to add.

GinoCanessa commented:
I've still only checked in the files for `src`, since it's cleaner - but I can check in the `dist` files as well.

jmandel commented on src/smart.ts:
```diff
@@ -134,6 +138,35 @@ function any(tasks: Task[]): Promise<any> {
+ function generatePKCECodes() {
+   var inputBytes:Buffer = jose.util.randomBytes(RECOMMENDED_CODE_VERIFIER_LENGTH);
+   var input:string = Array.from(inputBytes).map(function(val:number) { return PKCE_CHARSET[val % PKCE_CHARSET.length]; }).join('');
+   var codeVerifier:string = jose.util.base64url.encode(input);
```
This is base64urlencoding a string that's already (by  construction) a valid base64urlencoding. You're not getting the full entropy of your random bytes because of the `%` operation. Cleaner to do:

```
codeVerifier: string = jose.util.base64url.encode(jose.util.randomBytes(RECOMMENDED_CODE_VERIFIER_LENGTH))
```

jmandel commented on src/smart.ts:
```diff
@@ -134,6 +138,35 @@ function any(tasks: Task[]): Promise<any> {
+   var codeVerifier:string = jose.util.base64url.encode(input);
+   
+   jose.JWA.digest('SHA-256', codeVerifier).then(function(code:Buffer) {
+     return {
```
This returns from the `digest.then` closure, but ...  how does anyone ever get a hold of it?

jmandel commented on src/smart.ts:
```diff
@@ -14,6 +14,8 @@ import Client from "./Client";
 import { fhirclient } from "./types";
 
+//import * as jose from 'node-jose';
+var jose = require('node-jose');
```
Curious where the `//import ...` goes wrong?

jmandel commented:
> I didn't see a good place to add in tests for adding the `code_verifier` to a request - if there's a place where it's mocked out properly to enable that I'm happy to add.

The low hanging fruit might be to make sure at https://github.com/microsoft-healthcare-madison/client-js/blob/master/test/server.test.ts#L68-L82 that the following are covered:

* `smart1.authorize` has pkce ifSupported/disabled/enabled
  * mock discovery response has supported algorithms S256/nothing
    *  "authorize" request includes the challenge when it should, and doesn't include the challenge when it shouldn't

```ts
const pkceTests = [
  {"pkceMode": "required", "code_challenge_methods_supported": ["S256"], challengePresent: true]},
  {"pkceMode": "ifSupported", "code_challenge_methods_supported": ["S256"], challengePresent: true]},
  {"pkceMode": "disabled", "code_challenge_methods_supported": ["S256"], challengePresent: false]},
  {"pkceMode": "required", "code_challenge_methods_supported": [], challengePresent: false, throws: true]},
  {"pkceMode": "ifSupported", "code_challenge_methods_supported": [], challengePresent: false]},
  {"pkceMode": "disabled", "code_challenge_methods_supported": [], challengePresent: false]},
];

for (let t of pkceTests) {
  const pkceTestReq   = new HttpRequest("http://localhost/launch?launch=123&state=" + key);
  const pkceTestRes   = new HttpResponse();
  const pkceTestSmart = FHIR(pkceTestReq as any, pkceTestRes as any);

  mockServer.mock({
      headers: { "content-type": "application/json" },
      status: 200,
      body: {
          authorization_endpoint: mockUrl,
          token_endpoint: mockUrl,
          code_challenge_methods_supported: t.code_challenge_methods_supported
      }
  });
  try {
    await pkceTestSmart.authorize({
        iss        : mockUrl,
        scope      : "my_scope",
        clientId   : "my_client_id",
        redirectUri: "http://localhost/index",
        pkceMode: t.pkceMode
    });
    expect(t.throws).to.not.be.ok;
    const pkceTestAuthorizeUrl = new URL(pkceTestRes.headers.location);
    if (t.challengePresent) {
      expect(pkceTestAuthorizeUrl.searchParams.get("code_challenge_method")).to.equal("S256");
      expect(pkceTestAuthorizeUrl.searchParams.get("code_challenge")).to.have.lengthOf(43)      
    } else {
      expect(pkceTestAuthorizeUrl.searchParams.get("code_challenge_method")).to.be.undefined;
      expect(pkceTestAuthorizeUrl.searchParams.get("code_challenge")).to.be.undefined;
    }
  } catch (e) {
    expect(t.throws).to.be.ok;                
  }
}
```

GinoCanessa committed:
Improvements to pkce operation mode.
Better entropy for code verifier generation.

jmandel committed:
Use node-jose to create client assertions for authn

Add support for JWKS URL + tests

Move JWK signature tests to Smart.test.ts

Fix test names

Add token header + payload tests

Fix client_assertion_type

Fix auth method name for private_key_jwt

jmandel committed:
Await then return pkce code

jmandel committed:
Merge pull request #2 from microsoft-healthcare-madison/client-assertion-auth

Use node-jose to create client assertions for authn
